### PR TITLE
Proto: Adds support for add/remove validators in the sidechain

### PR DESCRIPTION
**Pull request by callistonianembrace**
## Depends

- [x] #184 
- [x] #185 
- [x] #187 

## Problem

Adding/removing validators is missing right now

## Solution

This PR adds protocol level support for submit addition and removal of validators

### CODE REVIEW

**Review by anchpop — APPROVED** [review]

**Comment by callistonianembrace** [comment]
@EduardoRFS Commits re-ordered

**Comment by EduardoRFS** [comment]
Fix the list of dependencies on the PR, and rebase this, the patch seems good tho.

**Comment by d4hines** [comment]
Quick scan of the PR looks fine to me.

However, I'm a bit concerned what this new property will do to consensus. In principle if we can verify the consensus of N validators, then we should be able to verify it when N changes (perhaps with certain constraints around when it's allowed to change). However, in practice I'm worried this will introduce more places for us to forget to check. The project is in a vulnerable state right now - not much unit or integration tests, let alone regression tests or formal verification. My two cents would be to hold off on this feature until we at least have a better specification for consensus (right now we don't even have a document that says what properties need to be checked, let alone a way of checking them).

**Comment by callistonianembrace** [comment]
This PR could definitely use more unit tests as @EduardoRFS has always been writing.

**Comment by d4hines** [comment]
> This PR could definitely use more unit tests as @EduardoRFS has always been writing.

That's not my issue. While unit tests always help to add confidence, I'm not doubting the correctness of your implementation of the feature. I'm doubting the safety of the feature in combination with everything else. This isn't something we can test because we haven't specified the global properties that the chain must have. For example, what constitute safety and liveness in our case? Different consensus algorithms guarantee different things, and I think we should spend some time specifying ours. Once we know what we're trying to guarantee, it will be easier to convince ourselves we indeed meet those guarantees.

File: protocol/protocol.re
Comment (callistonianembrace) [inline protocol/protocol.re:83]: Having `validator` parameter in the polyvar `Add_validator` doesn't seem useful. I think we can get rid of it

File: node/building_blocks.re
Comment (callistonianembrace) [inline node/building_blocks.re:84]: Should this be `>=`? Like when a node cli sends a Add/Remove operation to itself

File: node/building_blocks.re
Comment (EduardoRFS) [inline node/building_blocks.re:67]: The max network latency is closer to a second

File: node/building_blocks.re
Comment (EduardoRFS) [inline node/building_blocks.re:74]: Use `List.for_all`

File: node/building_blocks.re
Comment (EduardoRFS) [inline node/building_blocks.re:101]: This logic is duplicated from the case above and it's a really  important piece of domain logic, please abstract it in some meaningful function

File: node/building_blocks.re
Comment (EduardoRFS) [inline node/building_blocks.re:119]: This doesn't make sense here.

File: node/snapshots.re
Comment (EduardoRFS) [inline node/snapshots.re:14]: Why was this added to the snapshots?

File: bin/deku_node.re
Comment (EduardoRFS) [inline bin/deku_node.re:144]: This is not okay, you cannot send a timestamp from one validator to another.

File: protocol/operation.re
Comment (EduardoRFS) [inline protocol/operation.re:62]: This should not have a timestamp

File: protocol/operation.re
Comment (EduardoRFS) [inline protocol/operation.re:63]: Also I would say those two should be a single operation as every comment will always apply to both of them.

File: protocol/protocol.re
Comment (EduardoRFS) [reply protocol/protocol.re:89]: Agreed, if it is useful in the feature we can add it back.

File: protocol/protocol.re
Comment (EduardoRFS) [inline protocol/protocol.re:103]: This logic is duplicated from Add_validator

File: protocol/state.re
Comment (EduardoRFS) [inline protocol/state.re:16]: That's not right, this is part of the local consensus data, together with `last_state_root_update` and `last_applied_block_timestamp`

**Review by EduardoRFS — CHANGES_REQUESTED** [review]

File: node/building_blocks.re
Comment (callistonianembrace) [reply node/building_blocks.re:119]: Can you please elaborate?

File: protocol/protocol.re
Comment (callistonianembrace) [reply protocol/protocol.re]: Addressed in the next commit

File: node/building_blocks.re
Comment (callistonianembrace) [reply node/building_blocks.re]: Addressed by getting rid of it - timestamps are no longer a part of the operation design now.

File: node/building_blocks.re
Comment (callistonianembrace) [reply node/building_blocks.re]: Addressed in the next commit

File: node/building_blocks.re
Comment (callistonianembrace) [reply node/building_blocks.re]: Logic now simpler in the next commit now that timestamp is no longer  a part of operation and doesn't have to travel over the wire.

File: node/building_blocks.re
Comment (callistonianembrace) [reply node/building_blocks.re]: Moved to `Node.apply_block`, before `Protocol.apply_block`

File: node/snapshots.re
Comment (callistonianembrace) [reply node/snapshots.re:14]: So that all the validators have a common notion of `last_seen_membership_change_timestamp`

File: protocol/operation.re
Comment (callistonianembrace) [reply protocol/operation.re]: Addressed in the next commit by removing the timestamp from the operation.

File: protocol/protocol.re
Comment (callistonianembrace) [reply protocol/protocol.re]: abstracted into a single function. For now, Add/Remove are separate operations.

File: protocol/operation.re
Comment (callistonianembrace) [reply protocol/operation.re]: Let's revisit this again at the end.

File: protocol/state.re
Comment (callistonianembrace) [reply protocol/state.re:16]: I don't understand this comment - are you saying `last_seen_membership_change_timestamp` should be moved to node local state? `last_state_root_update` and `last_applied_block_timestamp` are currently a part of protocol state and not node - are you saying they should be moved too?

Comment (callistonianembrace) [reply protocol/state.re:16]: Thanks for explaining offline. Making the changes in the next commit

File: node/snapshots.re
Comment (callistonianembrace) [reply node/snapshots.re:14]: As explained offline, since this property is local consensus, ie values other nodes need not agree, I will be removing them from the snapshots.

File: node/state.re
Comment (EduardoRFS) [inline node/state.re:166]: This entire checking should not be here, this is a property of a signable block, you're not allowed to sign a block that contains a trusted validator change that you don't know, but you should still apply the block if 2/3 of the validators sign it.

**Review by EduardoRFS — CHANGES_REQUESTED** [review]

File: node/state.re
Comment (callistonianembrace) [reply node/state.re]: Fair point! How about moving this to `is_signable` ? Attempted this in the next incoming commit.

**Comment by EduardoRFS** [comment]
Looks okay to me, made some changes at https://github.com/marigold-dev/deku/pull/175, as soon as the CI is green I merge

**Review by EduardoRFS — APPROVED** [review]
deku vibing hard with the new feature

**Pull request merged**
